src/main/java/com/android/volley/CacheDispatcher.java

```diff
@@ -154,4 +172,109 @@ public void run() {
             }
         }
     }
+
+    private static class WaitingRequestHandler implements Request.NetworkRequestCompleteListener {
+	
+	/**
+	 * Staging area for requests that already have a duplicate request in flight.
+	 *
+	 * <ul>
+	 *     <li>containsKey(cacheKey) indicates that there is a request in flight for the given cache
+	 *          key.</li>
+	 *     <li>get(cacheKey) returns waiting requests for the given cache key. The in flight request
+	 *          is <em>not</em> contained in that list. Is null if no requests are staged.</li>
+	 * </ul>
+	 */
+	private final Map<String, Queue<Request<?>>> mWaitingRequests = new HashMap<>();
+
+	private final CacheDispatcher mCacheDispatcher;
+	
+	WaitingRequestHandler(CacheDispatcher cacheDispatcher) {
+	    mCacheDispatcher = cacheDispatcher;
+	}
+	
+	/** Request received a valid response that can be used by other waiting requests. */
+	@Override
+	public synchronized void onResponseReceived(Request<?> request, Response<?> response) {
```

Comment:
A key change here (since the last time I've looked at this) is that you are now holding a lock (via synchronized on the method) for all of the calls to postResponse. Previously it was not held during that time. I think this is probably reasonable given that a sane implementation will simply dispatch the response without actually waiting for the response to be handled (and that's the default behavior of ExecutorDelivery) but it's perhaps worth noting that a different delivery (i.e. ImmediateResponseDelivery, though this is only in the test code) would cause this lock to be held during response callback handling.

To summarize, no change requested here, but it'd be good to think through this angle one more time to make sure that there isn't a major increase in contention by holding a lock for each of these methods' entire bodies.

Reply:
I was going to comment on this as well: "Strive for *open calls* to avoid deadlock" being the best practice. In other words, don't hold a lock when calling other code.

In this case, the lock can be shrunk quite a bit:

[Hey. There are HARD tabs in this file. Is that cool?]

```
    Queue<Request<?>> waitingRequests;
    synchronized (this) {
        waitingRequests = mWaitingRequests.remove(cacheKey);
    }
    if (waitingRequests != null) {
        if (VolleyLog.DEBUG) {
            VolleyLog.v("Releasing %d waiting requests for cacheKey=%s.",
                    waitingRequests.size(), cacheKey);
        }
        // Process all queued up requests.
        for (Request<?> waiting : waitingRequests) {
            mCacheDispatcher.mDelivery.postResponse(waiting, response);
        }
    }
```

Reply:
Yeah, I think that's how it was in the initial PR, so we should probably move back to that unless there is a thread-safety issue. Just need to treat real carefully there :)

And for tabs, I'm not as used to GitHub's review tool so I haven't caught whitespace issues like that, but yes, we should be using spaces and not tabs. https://github.com/google/volley/issues/1 mentions integrating google-java-format into the workflow here which would resolve that more automatically FYI.

Reply:
I had that initially, but made it all synchronized for the inner class to avoid synchronized blocks. I thought that was part of the point of the nested class to have clearer synchronization by synchronizing the methods? But I can revert that. 
Apologies for the tabs, I was working on that in emacs at home, I'll have to check my configurations.

Reply:
@uhager sorry for the added confusion. The inner class separates concerns nicely, but the need for open calls adds a small complication. Thanks for addressing.